src/rotation.jl

```diff
@@ -0,0 +1,288 @@
+    return out
+end
+
+
+@kernel function imrotate_kernel_nearest!(out, arr, sinθ, cosθ, rotation_center, imax, jmax)
+    i, j, c, b = @index(Global, NTuple)
+
+    r(x...) = round(x..., RoundNearestTiesAway)
+    _, _, _, _, yrot_int, xrot_int = rotate_coordinates(sinθ, cosθ, i, j, rotation_center, r) 
+    if 1 ≤ yrot_int ≤ imax && 1 ≤ xrot_int ≤ jmax
+        @inbounds out[i, j, c, b] = arr[yrot_int, xrot_int, c, b]
+    end
+end
+
+
+@kernel function imrotate_kernel_bilinear!(out, arr, sinθ, cosθ, rotation_center, imax, jmax)
+    i, j, c, b = @index(Global, NTuple)
+    
+    yrot, xrot, yrot_f, xrot_f, yrot_int, xrot_int = rotate_coordinates(sinθ, cosθ, i, j, rotation_center, floor) 
+    if 1 ≤ yrot_int ≤ imax - 1 && 1 ≤ xrot_int ≤ jmax - 1 
+
+        ydiff, ydiff_1minus, xdiff, xdiff_1minus = 
+            bilinear_helper(yrot, xrot, yrot_f, xrot_f)
+        @inbounds out[i, j, c, b] = 
+            (   xdiff_1minus    * ydiff_1minus  * arr[yrot_int      , xrot_int      , c, b]
+             +  xdiff_1minus    * ydiff         * arr[yrot_int + 1  , xrot_int      , c, b]
+             +  xdiff           * ydiff_1minus  * arr[yrot_int      , xrot_int + 1  , c, b] 
+             +  xdiff           * ydiff         * arr[yrot_int + 1  , xrot_int + 1  , c, b])
+    end
+end
+
+
+@kernel function ∇imrotate_kernel_nearest!(out, arr, sinθ, cosθ, rotation_center, imax, jmax)
+    i, j, c, b = @index(Global, NTuple)
+
+    r(x...) = round(x..., RoundNearestTiesAway)
+    _, _, _, _, yrot_int, xrot_int = rotate_coordinates(sinθ, cosθ, i, j, rotation_center, r) 
+    if 1 ≤ yrot_int ≤ imax && 1 ≤ xrot_int ≤ jmax 
+        Atomix.@atomic out[yrot_int, xrot_int, c, b] += arr[i, j, c, b]
+    end
+end
+
+
+@kernel function ∇imrotate_kernel_bilinear!(out, arr, sinθ, cosθ, rotation_center, imax, jmax)
+    i, j, c, b = @index(Global, NTuple)
+
+    yrot, xrot, yrot_f, xrot_f, yrot_int, xrot_int = rotate_coordinates(sinθ, cosθ, i, j, rotation_center, floor) 
+    if 1 ≤ yrot_int ≤ imax - 1 && 1 ≤ xrot_int ≤ jmax - 1
+        o = arr[i, j, c, b]
+        ydiff, ydiff_1minus, xdiff, xdiff_1minus = 
+            bilinear_helper(yrot, xrot, yrot_f, xrot_f)
+        Atomix.@atomic out[yrot_int     ,   xrot_int    , c, b]  += xdiff_1minus    * ydiff_1minus * o
+        Atomix.@atomic out[yrot_int + 1 ,   xrot_int    , c, b]  += xdiff_1minus    * ydiff      * o
+        Atomix.@atomic out[yrot_int     ,   xrot_int + 1, c, b]  += xdiff           * ydiff_1minus * o
+        Atomix.@atomic out[yrot_int + 1 ,   xrot_int + 1, c, b]  += xdiff           * ydiff      * o
+    end
+end
+
+
+# is this rrule good? 
+# no @thunk and @unthunk
+function ChainRulesCore.rrule(::typeof(imrotate), arr::AbstractArray{T}, θ; 
+                              method=:bilinear, rotation_center=size(arr) .÷ 2 .+ 1) where T
+    res = imrotate(arr, θ; method, rotation_center)
+    function pb_rotate(dy)
+        ad = ∇imrotate(unthunk(dy), arr, θ; method, rotation_center)
```

Comment:
@mcabbott Is this rrule ok? I think thunking the output makes no sense here, right?

Reply:
Yeah I have no clue actually about it :smile:

Reply:
`unthunk(dy)` seems fine.

Why define `∇imrotate` at all though, and not just call `imrotate(arr, -θ)`?

Reply:
that is not the same under bilinear interpolation assumption, isn't it?
For nearest neighbour it is probably the same (excluding some rounding errors).

Reply:
```
julia> NNlib.∇imrotate(arr, arr, 0.1)
5×5×1×1 Array{Float64, 4}:
[:, :, 1, 1] =
 0.0         0.0        0.00994175  0.0848958  0.0
 0.0848958   0.90961    0.991056    0.90961    0.0
 0.00994175  0.991056   1.01799     0.991056   0.00994175
 0.0         0.90961    0.991056    0.90961    0.0848958
 0.0         0.0848958  0.00994175  0.0        0.0

julia> NNlib.imrotate(arr, -0.1)
5×5×1×1 Array{Float64, 4}:
[:, :, 1, 1] =
 0.0         0.0        0.00999167  0.0884453  0.0
 0.0884453   0.905162   1.0         0.905162   0.0
 0.00999167  1.0        1.0         1.0        0.00999167
 0.0         0.905162   1.0         0.905162   0.0884453
 0.0         0.0884453  0.00999167  0.0        0.0
``

Reply:
How does arr look like? And I think dy should be all ones to see the right gradient, not arr, correct?

Reply:
```
arr = ones((5,5,1,1)
arr[2:4, 2:4, 1,1] .= 1
```

Reply:
I think that makes sense because some pixels distribute their values over two pixels, so their rrule has 2s at exactly those pixels.

```
julia> NNlib.imrotate(arr, -0.8, method=:nearest)
5×5×1×1 Array{Float64, 4}:
[:, :, 1, 1] =
 0.0  0.0  1.0  0.0  0.0
 0.0  1.0  1.0  1.0  0.0
 1.0  1.0  1.0  1.0  1.0
 0.0  1.0  1.0  1.0  0.0
 0.0  0.0  1.0  0.0  0.0

julia> NNlib.∇imrotate(one.(arr), arr, -0.8, method=:nearest)
5×5×1×1 Array{Float64, 4}:
[:, :, 1, 1] =
 0.0  1.0  0.0  1.0  0.0
 1.0  2.0  1.0  2.0  1.0
 0.0  1.0  1.0  1.0  0.0
 1.0  2.0  1.0  2.0  1.0
 0.0  1.0  0.0  1.0  0.0
```

Reply:
I think it's ok to leave things as they are then. As long as it's correct and fast everything is ok :)

Reply:
Yeah I think it's fine and the `rrule` tests seem to work.